### PR TITLE
feat: add more fs cheatcodes

### LINKED ISSUE #4797: feat (cheatcode) allow to read folder
### Component

Forge

### Describe the feature you would like

I would like to get the list of file in a folder for my tool : [forge-deploy](https://github.com/wighawag/forge-deploy)

As of now `forge-deploy` do not need `--ffi` and while I could use `--ffi` to read the folder myself, I would like to avoid `--ffi` if possible

There is currently cheatcodes to read files knowing their path, but there is no such thing for folders

Proposal:
```
// Reads the directory and return the list of file/folder in it as a list of string
function readDir(string calldata) external returns (string[] memory);
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Fixes #4797

Added to `Cheats.sol`:

```solidity
struct DirEntry {
    string error_;
    string path;
    uint64 depth;
    bool isDir;
    bool isSymlink;
}

function createDir(string calldata, bool) external;

function removeDir(string calldata, bool) external;

function readDir(string calldata) external returns (DirEntry[] memory);
function readDir(string calldata, uint64) external returns (DirEntry[] memory);
function readDir(string calldata, uint64, bool) external returns (DirEntry[] memory);

function readLink(string calldata) external returns (string memory);
```

TODO: read dir test fails due to memory allocation overflow, I'm not sure why

### CODE REVIEW

File: testdata/cheats/Fs.t.sol
Comment (DaniPopes) [inline testdata/cheats/Fs.t.sol:1]: File.t.sol -> Fs.t.sol

File: testdata/cheats/Fs.t.sol [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        cheats.expectRevert(FOUNDRY_TOML_ACCESS_ERR);
+        cheats.writeFile("./../foundry.toml", "\nffi = true\n");
+    }
+
+    function testReadDir() public {
+        // TODO: Allocation overflow
+        // string memory path = "../testdata/fixtures/Dir";
+
+        // {
+        //     Cheats.DirEntry[] memory entries = cheats.readDir(path);
```
Comment (DaniPopes) [inline testdata/cheats/Fs.t.sol:144]: Just a single call fails with "allocation overflow", any ideas why?

File: evm/src/executor/abi/mod.rs [hunk 2]
```diff
@@ -134,6 +135,12 @@ abigen!(
         writeLine(string,string)
         closeFile(string)
         removeFile(string)
+        createDir(string, bool)
+        removeDir(string, bool)
+        readDir(string)(DirEntry[])
+        readDir(string, uint64)(DirEntry[])
+        readDir(string, uint64, bool)(DirEntry[])
+        readLink(string)(string)
```
Comment (mattsse) [inline evm/src/executor/abi/mod.rs:143]: very useful imo

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 3]
```diff
@@ -311,12 +307,116 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+/// - A parent of the given path doesn't exist and `recursive` is false.
+/// - `path` already exists and `recursive` is false.
+fn create_dir(state: &Cheatcodes, path: impl AsRef<Path>, recursive: bool) -> Result<Bytes, Bytes> {
+    let path =
+        state.config.ensure_path_allowed(path, FsAccessKind::Write).map_err(error::encode_error)?;
+    if recursive { fs::create_dir_all(path) } else { fs::create_dir(path) }
+        .map(|()| Bytes::new())
+        .map_err(error::encode_error)
+}
+
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:329]: I expect `fs` to grow eventually so would like to move all of them to new file `fs.rs`

File: testdata/cheats/Fs.t.sol (hunk 1 above)
Comment (mattsse) [reply testdata/cheats/Fs.t.sol:144]: remembered this issue:

https://github.com/foundry-rs/foundry/issues/4409

same error but turned out to be abi decoding related
but perhaps configured memory limit?

**Review by mattsse — CHANGES_REQUESTED** [review]
this is awesome!

codewise looks great, test look good.

suggestion re file layout, and failing test

File: testdata/cheats/Fs.t.sol (hunk 1 above)
Comment (DaniPopes) [reply testdata/cheats/Fs.t.sol]: Thanks, it was an ABI encoding error: 66723e4
Really weird that it throws a "memory overflow" tho

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 3 above)
Comment (DaniPopes) [reply evm/src/executor/inspector/cheatcodes/ext.rs]: done in f9eb363

File: testdata/cheats/Fs.t.sol (hunk 1 above)
Comment (mattsse) [reply testdata/cheats/Fs.t.sol]: yeh, perhaps our mapping of the internal solidity error is not entirely accurate

**Review by mattsse — APPROVED** [review]
this is great!

needs forge-std and book followups

**Pull request merged**